### PR TITLE
Clarify use of interface-ref throughout models.

### PR DESCRIPTION
**Pull request by robshakir**
```
 * (M) **/*.yang
   - Some implementations of the OpenConfig models today assume
     that some semantics are assigned to the key of a list of
     interfaces where the interface-ref type is used. This change
     clarifies that the interface and subinterface leaves are the
     authorative specification of which interface is being
     referenced, NOT the list key which may have arbitrary
     format.
```

### Change Scope

* Per-description, this change clarifies that the key of a list that contains an `interface-ref` type is not to be used to determine which interface is being referenced, and rather the `interface` and `subinterface` leaves are to be used.
* This change is backwards compatible, and simply clarifies an existing expectation.
* Tests to be added in [featureprofiles](https://github.com/openconfig/featureprofiles) to validate this behaviour.

* No platform implementations listed since this is a model clarification.

### CODE REVIEW

**Comment by dplore** [comment]
Reviewed in OC operators Feb 7, 2023 review with no objection.  

The key description change to read is https://github.com/openconfig/public/pull/802/files#diff-16872da203a8c932b7fb095e8886b9fe95b9ae3e0830aff9ab728b41f63fbfaaR223-R235

@earies @aredmon8551  @rgwilton @hobbydos @sulrich @rolandphung is the description clear?  Please ask clarifying questions.

**Comment by dplore** [comment]
Last call for comments, this will merge on Feb 21, 2023

**Comment by jsterne** [comment]
I'm a bit confused about this part: "Layer 2 interface is to be referenced, only the interface is specified."

How would a config look with the following:
- Physical port is using dot1q VLANs
- 5 VLANs are L3 interfaces
- 5 VLANs are L2 interfaces

Also: if there is a L2 interface (without an explicit subinterface), does the default 0 subinterface exist? In that case, what if a user explicitly configures subinterface 0 (and further, then a L3 NI binds to that subinterface? error?).

What if a L2 NI binds to a interface (only the 1st leafref), and then someone explcitly configures subinterface 55 on that same interface? error?  Only error if someone binds sub-if 55 to a L3 NI?

**Comment by robshakir** [comment]
You're right, some clarification here is required. I think the wording should be crisp about if this is an _entire_ L2 interface then only the interface is specified, if an L2 subinterface is required then the interface+subinterface are specified. Does this make sense?

I didn't quite follow your example: Is subinterface 0 in your example untagged? Is it L3? In the OpenConfig model, there is no need for an untagged interface to have a `subinterface 0`, AFAICS. Do you disagree?

**Comment by m-hesketh** [comment]
Is the intention of the PR that interface-ref is not mandatory to specify (which implies interface-id would be used when interface-ref is not present), but that if it is present it must be used instead of interface-id?

It feels like this change has the potential to open up a set of flexibility and complexity that isn't necessarily intended because the interface-id (as key) is immutable, whereas the interface-ref is not:
- Configuration is initially added with only interface-id present and interface-ref is added later
- interface-ref value can be changed by subsequent updates, whereas interface-id cannot

For cases where the interface is used internally as a key by an implementation, would it be reasonable for the implementation to require the interface-id and interface-ref to be consistent, and reject any configuration changes that introduce an inconsistency between the interface-id and interface-ref?

**Comment by robshakir** [comment]
It's actually the opposite -- it would be that `interface-ref` **is** mandatory to specify. We could mark `leaf interface` as mandatory (but can't mark `leaf subinterface` the same) inside the `interface-ref` container to indicate this more clearly.

**Comment by LimeHat** [comment]
> I didn't quite follow your example: Is subinterface 0 in your example untagged? Is it L3? In the OpenConfig model, there is no need for an untagged interface to have a `subinterface 0`, AFAICS. Do you disagree?

Rob, do you mean there's no need for **L2** untagged interfaces to have a subinterface 0?

It seems that the new text implies the existence of a subinterface for L3 scenarios (although not necessarily 0):

>  In contexts where a Layer 3 interface is to be referenced, both the
>         interface and subinterface leaves must be populated, as
>         Layer 3 configuration within the OpenConfig models is
>         associated with a subinterface.

**Comment by jsterne** [comment]
Hi Rob – taking a slightly different tack here. We might have different interpretations of the term “Layer 2 interface”.

Are you perhaps talking purely about situations where we are associating some object with an entire physical port? E.g. port level queues, LLDP, other L2CP protocols (things that have the scope of an entire port)? Maybe it is only those types of objects that can use the single leafref?

As a corollary: Is it correct that the top level binding between an NI and an interface must always use both leafrefs? (i.e. these ones here):

  +--rw network-instances
     +--rw network-instance* [name]
        +--rw interfaces
        |  +--rw interface* [id]
        |     +--rw id        -> ../config/id
        |     +--rw config
        |     |  +--rw id?                            string
        |     |  +--rw interface?                     -> /oc-if:interfaces/interface/name
        |     |  +--rw subinterface?                  -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index

(and likely similar with ACLs - must always use both leafrefs?)

**Comment by robshakir** [comment]
> > I didn't quite follow your example: Is subinterface 0 in your example untagged? Is it L3? In the OpenConfig model, there is no need for an untagged interface to have a `subinterface 0`, AFAICS. Do you disagree?
> 
> Rob, do you mean there's no need for **L2** untagged interfaces to have a subinterface 0?
> 
> It seems that the new text implies the existence of a subinterface for L3 scenarios (although not necessarily 0):
> 
> > In contexts where a Layer 3 interface is to be referenced, both the
> > interface and subinterface leaves must be populated, as
> > Layer 3 configuration within the OpenConfig models is
> > associated with a subinterface.

We have this document [here](https://github.com/openconfig/public/pull/573) that is being discussed on this.

From a first-principles perspective. OpenConfig `interface` entries do not have the ability to have addresses assigned to them, `subinterfaces` are for two purposes - either assigning L3 information or anywhere there is subdivision of an `interface` - for example, tagged routed VLAN interfaces (this cannot be specified on the `interface`).

For vendor models that allow:

```
interface FOO
  ipv4 1.2.3.4/32
```

then this does not conform to the way that OpenConfig models the interface, you have L3 information on the 'interface' in this model.

The document above standardises that this pseudo-config:

```
interface FOO
!
interface FOO.0
  ipv4 1.2.3.4/32
```

is equivalent to the above, since *both* are intended to have address `1.2.3.4/32` be assigned to OpenConfig `subinterface 0`.

Equally, the clarification above from a telemetry perspective says that the interface counters in this case for L3 counters that are not in `/interfaces/interface/state/counters` can be in /interfaces/interface[name=XXX]/subinterface/subinterface[index=0]/state/counters`.

OK, so now, what about L2-only interfaces? Here we have two cases:

1. Entire physical port is used in some context. In this scenario, there doesn't need to be L3 config (thus no need for a subinterface for this reason), and there is no subdivision of the interface. 
2. There is some subdivision of the interface. In this scenario there DOES need to be a subinterface to specify the encapsulation.

**Comment by robshakir** [comment]
> Hi Rob – taking a slightly different tack here. We might have different interpretations of the term “Layer 2 interface”.
> 
> Are you perhaps talking purely about situations where we are associating some object with an entire physical port? E.g. port level queues, LLDP, other L2CP protocols (things that have the scope of an entire port)? Maybe it is only those types of objects that can use the single leafref?

Yes, the entire physical port _only_ when it does not need L3 configuration is what I was referring to. This is what I meant by Layer 2 interface. To me, `/interfaces/interface[name=XX]/subinterface/subinterface[index=32]` with some VLAN encaps specified on it is a Layer 2 _subinterface_  but I recognise that there are different interpretations possible.

> 
> As a corollary: Is it correct that the top level binding between an NI and an interface must always use both leafrefs? (i.e. these ones here):

No, this isn't true when you are building a `L2P2P` NI that is associated with the entire physical port (port-mode PWE3).

> 
> ```
>   +--rw network-instances
>      +--rw network-instance* [name]
>         +--rw interfaces
>         |  +--rw interface* [id]
>         |     +--rw id        -> ../config/id
>         |     +--rw config
>         |     |  +--rw id?                            string
>         |     |  +--rw interface?                     -> /oc-if:interfaces/interface/name
>         |     |  +--rw subinterface?                  -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
> ```
> 
> (and likely similar with ACLs - must always use both leafrefs?)

Again, I don't think this is is the case if you are specifying an [`L2`](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-acl.html#ident-acl_l2) type ACL that is applied to an entire physical port.

**Comment by rgwilton** [comment]
Hi Rob S,

Okay, so this change seems to just be indicating that the interface-ref is the definition reference definition.  I presume that this means that implementations would be expected to reject configuration that hasn't specified a valid interface-ref?  I.e., it is not appropriate to create configuration with an inteface-id but no interface-ref and then expect it to be bound to an interface by configuring an interface-ref later?

Related to this, for implementations that allow for an interface-id to be entirely independent from the interface-ref, then does this mean that the interface-ref can change, and hence the applied config must be moved to a different underlying interface/sub-interface?

I presume that it is still acceptable for implementations that cannot support arbitrary interface-ids to constrain the interface-id to be defined consistently with the interface-ref?  But in these cases it is still required for the interface-ref to be populated.

Regards,
Rob

**Comment by dplore** [comment]
This PR will be a topic for the [OC public community meeting](https://docs.google.com/document/d/11sXUF2zujhbPs_bt-RFJhh2DcIbEvnOpd9GypE-u3cc/edit#).

**Comment by jsterne** [comment]
Hi all,

Overall I'm still not keen on having this exception for L2 interfaces. Why not stay consistent and always require a subinterface for either L3 or L2? 

The data model would support the following. Is there any thoughts that this is or isn't supposed to be supported (or just leaving that as vendor specific)?

##########
Scenario A:
##########
NI #1:
- interface-ref binds to interface "foo" (no subinterface binding)
NI #2:
- interface-ref binds to interface "foo", subinterface 23.
Interface "foo" 
- dot1q encap
- subinterface 23:  associated with VLAN 100

All traffic on VLAN 100 (in port "foo") would be part of subinterface 23 and NI #2 
"All the rest of the traffic" arriving on port "foo" would be part of NI #1. This would include any other tagged traffic, untagged traffic or traffic with VLAN=0 (i.e. prio tag only)

File: release/models/policy-forwarding/openconfig-pf-interfaces.yang [hunk 1]
```diff
@@ -151,9 +162,12 @@ submodule openconfig-pf-interfaces {
         path "../../../../policies/policy/config/policy-id";
       }
       description
-        "Apply the specific VRF selection policy on the interface. The referenced policy
-        MUST be of type VRF_SELECTION_POLICY. The VRF selection policy may coexist with
-        a policy-forwarding policy.
+        "Apply the specific VRF selection policy on the interface. 
+        Interface where the policy has to be binded REQUIRES to be defined
+        under the interface-ref container using the Interface and Sub-interface leaves.
```
Comment (dplore) [inline release/models/policy-forwarding/openconfig-pf-interfaces.yang:167]:
```suggestion
        The referenced Interface must be resolved using the Interface 
        and Sub-interface leaves.
```

**Comment by dplore** [comment]
> Hi Rob S,
> 
> Okay, so this change seems to just be indicating that the interface-ref is the definition reference definition. I presume that this means that implementations would be expected to reject configuration that hasn't specified a valid interface-ref? I.e., it is not appropriate to create configuration with an inteface-id but no interface-ref and then expect it to be bound to an interface by configuring an interface-ref later?

Perhaps this is a clear way to state the rule for resolving leaf-ref to interfaces:   A leaf-ref  interface must be resolved using `/interfaces/interface/name` and `/interfaces/interface/subinterface/index` (where appropriate) which are the canonical names for an interface.  `interface-id` is to be treated as an index into the interface list and should not parsed for content to resolve interface references.

> Related to this, for implementations that allow for an interface-id to be entirely independent from the interface-ref, then does this mean that the interface-ref can change, and hence the applied config must be moved to a different underlying interface/sub-interface?

Yes, this is true.  

> I presume that it is still acceptable for implementations that cannot support arbitrary interface-ids to constrain the interface-id to be defined consistently with the interface-ref? But in these cases it is still required for the interface-ref to be populated.

Yes, this is also true.

File: release/models/platform/openconfig-platform-port.yang [hunk 2]
```diff
@@ -24,7 +24,14 @@ module openconfig-platform-port {
     "This module defines data related to PORT components in the
     openconfig-platform model";
 
-  oc-ext:openconfig-version "0.4.2";
+  oc-ext:openconfig-version "1.0.0";
```
[the remark this answers is not in the document]
Comment (dplore) [reply release/models/platform/openconfig-platform-port.yang]: Undid the v1.0 promotion here.   There is no special reason to promote in this PR.  We can promote this model to v1.0 in a future PR.

**Comment by dplore** [comment]
I'll fix the CI checks and then make this last call.

**Comment by robshakir** [comment]
> Hi all,
> 
> Overall I'm still not keen on having this exception for L2 interfaces. Why not stay consistent and always require a subinterface for either L3 or L2?

I’m open to this discussion — it represents a difference to what we’ve thought before, but I agree that this is definitely something that could be defined. 

I assume the proposal would be that we would say that from a configuration perspective one need always create a `subinterface` for any interface. 

We have some L2 configuration on a subinterface already. For example, through the [`match`](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-interfaces.html#interfaces-interface-subinterfaces-subinterface-vlan-match) as a way to a sub-interface can match a specific set of VLANs (see more notes on this later in this reply).

Today, we have two complexities that would not allow for this:

 * `routed-vlan` and `tunnel` interfaces that themselves are not subdivisible do not have subinterfaces. We would need to change this and require that they have such configuration in all scenarios (this would mean rethinking modelling of both).
 * we have `switched-vlan` that specifies allowed VLANs on a trunk of a switched device, or access port to VLAN mapping, that is outside of a subinterface. We would probably have to think about bringing it into the subinterface context if L2 interfaces require subinterfaces.

These are changes that we could make, but we need to go into this eyes-open to the impact on the models.

> 
> The data model would support the following. Is there any thoughts that this is or isn't supposed to be supported (or just leaving that as vendor specific)?
> 
> ########## Scenario A: ########## NI 1:
> 
> * interface-ref binds to interface "foo" (no subinterface binding)
> 
> NI 2:
> 
> * interface-ref binds to interface "foo", subinterface 23
> 
> Interface "foo":
> 
> * dot1q encap
> * subinterface 23:  associated with VLAN 100
> 
> All traffic on VLAN 100 (in port "foo") would be part of subinterface 23 and NI 2 "All the rest of the traffic" arriving on port "foo" would be part of NI 1. This would include any other tagged traffic, untagged traffic or traffic with VLAN=0 (i.e. prio tag only)

I think we need to make sure that this is clearly defined :-) Today, I think we have not defined this to say that on a tagged interface there is a “longest match” of the VLANs. There was some discussion of this on the addition of `match`, but I think today we say that there is an explicit wildcarding to be able to do what you specify above (i.e., we would need `/interfaces/interface[name=foo]/subinterfaces/subinterface[index=0]/vlan/match/…` to specify that there is a wildcard match, which is then bound to NI1.

**Comment by robshakir** [comment]
> > I presume that it is still acceptable for implementations that cannot support arbitrary interface-ids to constrain the interface-id to be defined consistently with the interface-ref? But in these cases it is still required for the interface-ref to be populated.
> 
> Yes, this is also true.

I think this is true, but suboptimal. It should be possible to name the key anything that one wants (there are no such constraints described in the model). It is OK to require a specific format **if and only if** the `interface` and `subinterface` leaves are used to resolve the interface that is being referenced.

**Comment by robshakir** [comment]
> I'll fix the CI checks and then make this last call.

I think here we need to ensure that we answer how L2 untagged interfaces are referenced before making this last call.

**Comment by jsterne** [comment]
> There was some discussion of this on the addition of match, but I think today we say that there is an explicit wildcarding to be able to do what you specify above (i.e., we would need /interfaces/interface[name=foo]/subinterfaces/subinterface[index=0]/vlan/match/… to specify that there is a wildcard match, which is then bound to NI1.

Yes - having some sort of explicit config that says "all the rest of the VLANs that aren't already specified/matched" would probably be good.

**Comment by jsterne** [comment]
> I assume the proposal would be that we would say that from a configuration perspective one need always create a subinterface for any interface.

Yes - I think that may be the case (at least before it is bound to an NI). The top level binding to an NI would always use the 2 leafrefs.

That doesn't necessarily mean that all bindings to an interface everywhere in the model definitely need both leafrefs though. Even though a L2 subinterface would exist, maybe objects that have the scope of an entire port would just use the single leafref binding (E.g. port level queues, LLDP, other L2CP protocols).

**Comment by dplore** [comment]
I'd like to propose that we take the L2 sub interface clarifications into a separate PR.  If the change does require subinterface on every interface, that is a major, breaking change.     Right now, this clarification for leaf-ref adds clarity to the existing model where we have seen issues in featureprofiles testing and is non-breaking.

**Comment by rgwilton** [comment]
> I'd like to propose that we take the L2 sub interface clarifications into a separate PR. If the change does require subinterface on every interface, that is a major, breaking change. Right now, this clarification for leaf-ref adds clarity to the existing model where we have seen issues in featureprofiles testing and is non-breaking.

I also think that we need to be very careful that we are not talking cross purposes where we are referring to L2 interfaces.  E.g., historically, I see a difference at configuration level between L2 trunking and VLAN switching vs a sub-interface service based model (e.g., where a trunk is divided up into sub-interfaces, some of which terminate to L3, and others are switched at L2 via P2P xconnects or L2 bridging instances).

I think that we also need to think carefully about whether we want to force every interface, specifically tunnels and other logical interfaces to always have sub-interfaces.  This has a potential of negatively impacting device scalability depending on how devices implement them, although it does have the potential benefit of simplifying other parts of the model.

But for me, I think that most of the interface handling complexity in the OC model comes from having sub-interface not be interfaces.  E.g., moving the ports out to a separate list and having all interfaces/sub-interfaces as a single list under /interfaces/ would probably simplify a lot of the interface referencing discussions/issues (and align tunnels and sub-interfaces to be handled the same way).  Of course, that would be an absolutely huge breaking change to make to the Open Config models, which I appreciate is presumably not realistic or desirable at this stage ....

But generally, I'm supportive of completing the current change/clarification, and moving the discussion on L2 interfaces to a separate issue.

**Review by dplore — APPROVED** [review]
LGTM.  This will merge on Apr 25th.  I will correct the model version numbers at that time.

**Pull request merged**